Data Analyst: Andrei Berner Arroyo Tanta

**Project with Numpy 1: Multidimensional Inventory & Quality Optimization System

**Business Application: This project focuses on the operational efficiency of a medium-sized Italian wine producer. In the wine industry, managing vast quantities of liquid assets and bottled stock requires high-performance tracking to prevent capital stagnation and ensure quality standards.

-Business Need: Centralize the management of stock across various tanks and batches while monitoring chemical properties like Alcohol By Volume (ABV).

-Area of Benefit: Supply Chain, Quality Control, and Financial Planning.

-Decision Support: Enables automated mass replenishment of active tanks and provides instant financial forecasting for inflation-adjusted pricing.

**Executive Summary

-Asset Overview: The system manages a total volume of 562,699.38 liters with a market value of €1,679,111.38.

-Quality Portfolio: Identified 72 premium products with a high-strength alcohol content exceeding 14%, alongsi-de a stable average ABV of 13.42% across the catalog.

-Economic Impact: A simulated 10% inflation rate showed an average cost increase from €2.63 to €2.89. The system confirmed that entry-level products remained competitive with a minimum price floor shift to only €1.21.

-Technical Efficiency: Utilized optimized data types like uint16 and float32 to minimize memory footprint and employed vectorized "views" for lightning-fast calculations without data duplication.

In [1]:
#Although this project is about NumPy, we will load pandas exclusively for the operation of loading data from a dataset.
import numpy as np
import pandas as pd

# Suppress scientific notation for better readability of financial and technical data
np.set_printoptions(suppress=True, precision=2)

In [2]:
#Loading raw data
dataset_path = "stock_wine.xls"
df = pd.read_csv(dataset_path)

#Defining our multidimensional array with an optimized dtype
#Initializing with np.empty to pre-allocate memory
num_rows, num_cols = df.shape
stock = np.empty((num_rows, num_cols), dtype=np.float32)

#Transfering data from Pandas to NumPy
stock[:] = df.to_numpy()

print(f"Array Shape: {stock.shape}")
print(f"Data Type: {stock.dtype}")
print(f"Memory Usage: {stock.nbytes} bytes")

Array Shape: (252, 10)
Data Type: float32
Memory Usage: 10080 bytes


In [3]:
#To keep the code readable and performant, we use NumPy Slicing.

In [4]:
#Mapping columns to descriptive variables (Vectorized Views)
reg_codes = stock[:, 0].astype(np.uint16) # Casting to uint16 for optimization
product_type = stock[:, 1]
product_status = stock[:, 2]
batch_id = stock[:, 3]
tank_id = stock[:, 4]
volume_liters = stock[:, 5]
bottle_count = stock[:, 6]
alcohol_content = stock[:, 7]
unit_cost = stock[:, 8]
inventory_value = stock[:, 9]

In [5]:
#Inventory Diagnosis (Vectorized Aggregations)
#Aggregate Metrics
total_liters  = np.sum(volume_liters)
total_bottles = np.sum(bottle_count)
total_market_value = np.sum(inventory_value)

print(f"--- Inventory Report ---")
print(f"Total Volume: {total_liters:,.2f} L")
print(f"Bottled Stock: {total_bottles:,.0f} units")
print(f"Total Asset Value: €{total_market_value:,.2f}")

--- Inventory Report ---
Total Volume: 562,699.38 L
Bottled Stock: 296,191 units
Total Asset Value: €1,679,111.38


Our warehouse has a solid infrastructure, managing a total volume of 562,699.38 liters of wine valued at €1,679,111.38. This represents a significant current asset that requires efficient risk management. The inventory is divided into 296,191 bottled units, which means we have immediate response capacity for commercial distribution.

In [6]:
#Quality Analysis & Boolean Indexing
# 1. Alcohol content analysis
avg_alcohol = np.mean(alcohol_content)

#High-strength filter (Boolean Indexing)
high_alcohol_mask = alcohol_content > 14.0
high_alcohol_count = np.count_nonzero(high_alcohol_mask)

#Calculate Pure Alcohol Volume per unit (Vectorized Arithmetic)
#Formula: (Volume * ABV) / 100
pure_alcohol_units = np.around((volume_liters * alcohol_content) / 100, decimals=2)

print(f"--- Inventory Report ---")
print(f"Average ABV: {avg_alcohol:.2f}%")
print(f"Premium/High-Strength Wines (>14%): {high_alcohol_count} products")

--- Inventory Report ---
Average ABV: 13.42%
Premium/High-Strength Wines (>14%): 72 products


In [7]:
#Economic Modeling & Bulk Replenishment
#Inflation Simulation: Increase unit cost by 10% across the board
inflation_rate = 1.10
inflated_costs = unit_cost * inflation_rate

#Bulk Replenishment: Simulate adding 500L to all tanks in 'Active' status
#Let's assume Status 1.0 = 'Active'
replenishment_mask = (product_status == 1.0)
volume_liters[replenishment_mask] += 500 

#Compare Cost Changes
min_orig_cost = np.min(unit_cost)
min_new_cost = np.min(inflated_costs)

print(f"--- Inventory Report ---")
print(f"Original Avg Cost: €{np.mean(unit_cost):.2f}")
print(f"Post-Inflation Avg Cost: €{np.mean(inflated_costs):.2f}")
print(f"Lowest Price Point Change: €{min_orig_cost:.2f} to €{min_new_cost:.2f}")

--- Inventory Report ---
Original Avg Cost: €2.63
Post-Inflation Avg Cost: €2.89
Lowest Price Point Change: €1.10 to €1.21


Our current wine portfolio reflects a balanced and strategically positioned product mix. The average alcohol content across the catalog stands at 13.42% ABV, and we also maintain 72 premium or high‑strength wines (above 14% ABV).

From a cost‑management perspective, the portfolio shows a controlled inflation impact. The original average cost per unit was €2.63, while the post‑inflation adjusted average now stands at €2.89.

Additionally, the lowest price point in the catalog increased from €1.10 to €1.21, a change that remains within acceptable thresholds for entry‑level products and preserves competitiveness in the lower‑cost segment.

In [8]:
#Dynamic Data Concatenation
#Loading additional inventory data
df_extra = pd.read_csv("stock_wine_2.csv") #Replace with name of file
extra_stock = df_extra.to_numpy(dtype=np.float32)

#Vertically stacking the new production rows to the existing stock
#axis=0 ensures we add new rows, not new columns
updated_stock = np.concatenate((stock, extra_stock), axis=0)

print(f"Initial Inventory Rows: {stock.shape[0]}")
print(f"New Inventory Rows: {updated_stock.shape[0]}")

FileNotFoundError: [Errno 2] No such file or directory: 'stock_wine_2.csv'